apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/InsertOptions.cs

```diff
@@ -30,12 +30,48 @@ public sealed class InsertOptions
         /// </summary>
         public bool? AllowUnknownFields { get; set; }
 
+        /// <summary>
+        /// Insert all valid rows of a request, even if invalid rows exist. The
+        /// default value is false, which causes the entire request to fail if any invalid
+        /// rows exist.
+        /// Use in combination with <see cref="ThrowOnSkippedInvalidRows"/> to obtain
+        /// information about the rows that were skipped.
+        /// </summary>
+        public bool? SkipInvalidRows { get; set; }
+
+        /// <summary>
+        /// If <see cref="SkipInvalidRows"/> is set to <c>true</c>, then you can use
+        /// this option to control whether the insertion silently skips the invalid rows
+        /// or if it throws a <see cref="GoogleApiException"/> with information on each of the
+        /// rows that failed. Either way, as long as <see cref="SkipInvalidRows"/> is set to
+        /// <c>true</c> all valid rows will be inserted.
+        /// The default value is false.
+        /// </summary>
+        public bool ThrowOnSkippedInvalidRows { get; set; }
```

Comment:
Hmm. I'm not sure about this. It will make it hard to tell the difference between "Yes, really the server threw an error - we haven't inserted" and "No, everything's been inserted". It's a shame we don't return the insertion response, to be honest.

Do we know if other languages have this?

How would you feel about not doing this part yet, but starting a feature request for it and getting feedback from the BQ team?

Reply:
Yep, I thought about this as well, agree that returning the response would have been better.

I did it like this in the end because I found the original feature request issue (I'll share that internally), and the main use case they were trying to solve was for:
- Skip bad rows knowing why they had been skipped
- For the ones that had been skipped because of unknown fields, attempt a new insert with `AllowUnknownFields` set to true, which is always silent.
True that would have been solved by always throwing, but I think that by default I would expect not throwing anything when `SkipInvalidRows` is set to true.

I'm fine to move this to another PR and chat with the BQ team about it. we just need to make a decision about what to do when SkipInvalidRows is set to true. I'm torn:

- I **like** skipping silently, but that doesn't cover use cases in which knowing what was skipped and why is important, so maybe the _right_ thing is to throw if skipped.

And I just looked at Java and they are returning an [`InsertAllResponse`](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/InsertAllResponse.java) which can contain errors for skipped rows.

Reply:
How about we go without the extra property, temporarily making it hard to detect this, but then change the return type in our "next major version" bump?

Reply:
Yes, very nice!!

> temporarily making it hard to detect this

So, without the extra property, should we skip rows silently (and it's hard to detect what was skipped) or throw when skipped (and it's hard to distinguish skipping erros from something else failing)? I prefer being silent but I'm fine either way.

Reply:
If possible, I think we should be silent, but it's not obvious how feasible that is. We don't want to skip *all* errors just because we've asked to skip invalid rows. Maybe the table isn't present at all! I think we probably need to think about that more. Presumably if we leave all other code (e.g. HandleInsertAllResponse) as it was, it will throw. I suspect that's okay, and what we *actually* need in the next version is an option of "Don't throw on errors; return the result to me instead."